File: WCCBO2/Shizuku2Lib/OccupantCommunicator.cs
using System.IO.BACnet;

namespace Shizuku2.BACnet
{

  /// <summary>Shizuku2のOccupantモニタとの通信ユーティリティクラス</summary>
  public class OccupantCommunicator : PresentValueReadWriter
  {

    #region 定数宣言

    /// <summary>Device ID</summary>
    public const uint OCCUPANTMONITOR_DEVICE_ID = 5;

    /// <summary>排他的ポート番号</summary>
    public const int OCCUPANTMONITOR_EXCLUSIVE_PORT = 0xBAC0 + (int)OCCUPANTMONITOR_DEVICE_ID;

    /// <summary>OccupantモニタのBACnetアドレス</summary>
    private readonly BacnetAddress bacAddress;

    #endregion

    #region 列挙型

    /// <summary>項目</summary>
    private enum memberNumber
    {
      /// <summary>執務者の数</summary>
      OccupantNumber = 1,
      /// <summary>在室状況</summary>
      Availability = 2,
      /// <summary>温冷感</summary>
      ThermalSensation = 3,
      /// <summary>着衣量</summary>
      ClothingIndex = 4,
      /// <summary>熱的な不満足者率</summary>
      Dissatisfied_Thermal,
      /// <summary>ドラフトによる不満足者率</summary>
      Dissatisfied_Draft,
      /// <summary>上下温度分布による不満足者率</summary>
      Dissatisfied_VerticalTemp
    }

    /// <summary>テナント</summary>
    public enum Tenant
    {
      /// <summary>南テナント</summary>
      South = 1,
      /// <summary>北テナント</summary>
      North = 2
    }

    /// <summary>温冷感申告値</summary>
    public enum ThermalSensation
    {
      /// <summary>Cold</summary>
      Cold = -3,
      /// <summary>Cool</summary>
      Cool = -2,
      /// <summary>Slightly Cool</summary>
      SlightlyCool = -1,
      /// <summary>Neutral</summary>
      Neutral = 0,
      /// <summary>Slightly Warm</summary>
      SlightlyWarm = 1,
      /// <summary>Warm</summary>
      Warm = 2,
      /// <summary>Hot</summary>
      Hot = 3
    }

    #endregion

    #region コンストラクタ

    /// <summary>インスタンスを初期化する</summary>
    /// <param name="id">通信に使うBACnet DeviceのID</param>
    /// <param name="name">通信に使うBACnet Deviceの名前</param>
    /// <param name="ipAddress">エミュレータのIPアドレス（「xxx.xxx.xxx.xxx」の形式）</param>
    public OccupantCommunicator(uint id, string name = "anoymous device", string ipAddress = "127.0.0.1")
      : base(id, name)
    {
      bacAddress = new BacnetAddress(BacnetAddressTypes.IP, ipAddress + ":" + OCCUPANTMONITOR_EXCLUSIVE_PORT.ToString());
    }

    #endregion

    #region テナント・ゾーン別

    /// <summary>在室している執務者数を取得する</summary>
    /// <param name="tenant">テナント</param>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>在室している執務者数</returns>
    public int GetOccupantNumber(Tenant tenant, out bool succeeded)
    {
      uint instNum = (uint)(10000 * (int)tenant + (int)memberNumber.OccupantNumber);
      return ReadPresentValue<int>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_INPUT, instNum, out succeeded);
    }

    /// <summary>ゾーンに在室している執務者数を取得する</summary>
    /// <param name="tenant">テナント</param>
    /// <param name="zoneNumber">ゾーン番号</param>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>ゾーンに在室している執務者数</returns>
    public int GetOccupantNumber(Tenant tenant, int zoneNumber, out bool succeeded)
    {
      uint instNum = (uint)(10000 * (int)tenant + 1000 * zoneNumber + (int)memberNumber.OccupantNumber);
      return ReadPresentValue<int>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_INPUT, instNum, out succeeded);
    }

    /// <summary>ゾーンに在室している執務者の平均温冷感申告値を取得する</summary>
    /// <param name="tenant">テナント</param>
    /// <param name="zoneNumber">ゾーン番号</param>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>ゾーンに在室している執務者の平均温冷感申告値</returns>
    public float GetAveragedThermalSensation(Tenant tenant, int zoneNumber, out bool succeeded)
    {
      uint instNum = (uint)(10000 * (int)tenant + 1000 * zoneNumber + (int)memberNumber.ThermalSensation);
      return ReadPresentValue<float>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_INPUT, instNum, out succeeded);
    }

    /// <summary>ゾーンに在室している執務者の平均着衣量を取得する</summary>
    /// <param name="tenant">テナント</param>
    /// <param name="zoneNumber">ゾーン番号</param>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>ゾーンに在室している執務者の平均着衣量</returns>
    public float GetAveragedClothingIndex(Tenant tenant, int zoneNumber, out bool succeeded)
    {
      uint instNum = (uint)(10000 * (int)tenant + 1000 * zoneNumber + (int)memberNumber.ClothingIndex);
      return ReadPresentValue<float>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_INPUT, instNum, out succeeded);
    }

    /// <summary>ゾーンに在室している執務者の温熱環境に対する不満足者率を取得する</summary>
    /// <param name="tenant">テナント</param>
    /// <param name="zoneNumber">ゾーン番号</param>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>ゾーンに在室している執務者の温熱環境に対する不満足者率</returns>
    public float GetThermallyDissatisfiedRate(Tenant tenant, int zoneNumber, out bool succeeded)
    {
      uint instNum = (uint)(10000 * (int)tenant + 1000 * zoneNumber + (int)memberNumber.Dissatisfied_Thermal);
      return ReadPresentValue<float>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_INPUT, instNum, out succeeded);
    }

    /// <summary>ゾーンに在室している執務者のドラフトに対する不満足者率を取得する</summary>
    /// <param name="tenant">テナント</param>
    /// <param name="zoneNumber">ゾーン番号</param>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>ゾーンに在室している執務者のドラフトに対する不満足者率</returns>
    public float GetDissatisfiedRateCausedByDraft(Tenant tenant, int zoneNumber, out bool succeeded)
    {
      uint instNum = (uint)(10000 * (int)tenant + 1000 * zoneNumber + (int)memberNumber.Dissatisfied_Draft);
      return ReadPresentValue<float>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_INPUT, instNum, out succeeded);
    }

    /// <summary>ゾーンに在室している執務者の上下温度分布に対する不満足者率を取得する</summary>
    /// <param name="tenant">テナント</param>
    /// <param name="zoneNumber">ゾーン番号</param>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>ゾーンに在室している執務者の上下温度分布に対する不満足者率</returns>
    public float GetDissatisfiedRateCausedByVerticalTemperatureDistribution(Tenant tenant, int zoneNumber, out bool succeeded)
    {
      uint instNum = (uint)(10000 * (int)tenant + 1000 * zoneNumber + (int)memberNumber.Dissatisfied_VerticalTemp);
      return ReadPresentValue<float>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_INPUT, instNum, out succeeded);
    }

    #endregion

    #region 執務者別

    /// <summary>在室しているか否かを取得する</summary>
    /// <param name="tenant">テナント</param>
    /// <param name="occupantIndex">執務者番号（1～）</param>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>在室しているか否か</returns>
    public bool IsOccupantStayInOffice(Tenant tenant, int occupantIndex, out bool succeeded)
    {
      uint instNum = (uint)(10000 * (int)tenant + 10 * occupantIndex + (int)memberNumber.Availability);
      return 1 == ReadPresentValue<uint>
        (bacAddress, BacnetObjectTypes.OBJECT_BINARY_INPUT, instNum, out succeeded);
    }

    /// <summary>温冷感を取得する</summary>
    /// <param name="tenant">テナント</param>
    /// <param name="occupantIndex">執務者番号（1～）</param>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>温冷感</returns>
    public ThermalSensation GetThermalSensation(Tenant tenant, int occupantIndex, out bool succeeded)
    {
      uint instNum = (uint)(10000 * (int)tenant + 10 * occupantIndex + (int)memberNumber.ThermalSensation);
      return convertVote(ReadPresentValue<int>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_INPUT, instNum, out succeeded));
    }

    /// <summary>着衣量を取得する</summary>
    /// <param name="tenant">テナント</param>
    /// <param name="occupantIndex">執務者番号（1～）</param>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>着衣量</returns>
    public float GetClothingIndex(Tenant tenant, int occupantIndex, out bool succeeded)
    {
      uint instNum = (uint)(10000 * (int)tenant + 10 * occupantIndex + (int)memberNumber.ClothingIndex);
      return ReadPresentValue<float>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_INPUT, instNum, out succeeded);
    }

    #endregion

    #region privateメソッド

    private ThermalSensation convertVote(int vote)
    {
      switch (vote)
      {
        case -3:
          return ThermalSensation.Cold;
        case -2:
          return ThermalSensation.Cool;
        case -1:
          return ThermalSensation.SlightlyCool;
        case 0:
          return ThermalSensation.Neutral;
        case 1:
          return ThermalSensation.SlightlyWarm;
        case 2:
          return ThermalSensation.Warm;
        case 3:
          return ThermalSensation.Hot;
        default:
          return ThermalSensation.Neutral;
      }
    }

    #endregion

  }
}

File: Samples/python/bacpypes3/OccupantCommunicator.py
import PresentValueReadWriter
import time
import asyncio

from enum import Enum

class OccupantCommunicator(PresentValueReadWriter.PresentValueReadWriter):

# region 定数宣言

    OCCUPANTMONITOR_DEVICE_ID = 5

    OCCUPANTMONITOR_EXCLUSIVE_PORT = 0xBAC0 + OCCUPANTMONITOR_DEVICE_ID

# endregion

# region 列挙型定義

    class _member(Enum):
        # 執務者の数
        OccupantNumber = 1
        # 在室状況
        Availability = 2
        # 温冷感
        ThermalSensation = 3
        # 着衣量
        ClothingIndex = 4
        # 熱的な不満足者率
        Dissatisfied_Thermal = 5
        # ドラフトによる不満足者率
        Dissatisfied_Draft = 6
        # 上下温度分布による不満足者率
        Dissatisfied_VerticalTemp = 7

    class Tenant(Enum):
        # 南テナント
        South = 1
        # 北テナント
        North = 2

    class ThermalSensation(Enum):
        # Cold
        Cold = -3
        # Cool
        Cool = -2
        # Slightly Cool
        SlightlyCool = -1
        # Neutral
        Neutral = 0
        # Slightly Warm
        SlightlyWarm = 1
        # Warm
        Warm = 2
        # Hot
        Hot = 3

# endregion

# region コンストラクタ

    def __init__(self, id, name='occComm', device_ip='127.0.0.1', emulator_ip='127.0.0.1', time_out_sec=1.0):
        """インスタンスを初期化する

        Args:
            id (int): 通信用のDeviceのID
            name (str): 通信用のDeviceの名前
            device_ip (str): 通信に使うDeviceのIP Address（xxx.xxx.xxx.xxx）
            emulator_ip (str): エミュレータのIP Address（xxx.xxx.xxx.xxx）
        """
        super().__init__(id, name, device_ip, emulator_ip, time_out_sec)
        self.target_ip = emulator_ip + ':' + str(self.OCCUPANTMONITOR_EXCLUSIVE_PORT)

# endregion

# region テナント・ゾーン別

    async def get_occupant_number(self, tenant):
        """在室している執務者数を取得する
        Args:
            tenant (Tenant): テナント
        Returns:
            list: 読み取り成功の真偽,在室している執務者数
        """
        inst = 'analogInput:' + str(10000 * int(tenant.value) + self._member.OccupantNumber.value)       
        return await self.read_present_value(self.target_ip,inst)


    async def get_zone_occupant_number(self, tenant, zone_number):
        """ゾーンに在室している執務者数を取得する
        Args:
            tenant (Tenant): テナント
            zone_number (Unsigned): ゾーン番号
        Returns:
            list: 読み取り成功の真偽,ゾーンに在室している執務者数
        """
        inst = 'analogInput:' + str(10000 * int(tenant.value) + 1000 * zone_number + self._member.OccupantNumber.value)       
        return await self.read_present_value(self.target_ip,inst)


    async def get_averaged_thermal_sensation(self, tenant, zone_number):
        """ゾーンに在室している執務者数の平均温冷感を取得する
        Args:
            tenant (Tenant): テナント
            zone_number (Unsigned): ゾーン番号（1~9）
        Returns:
            list: 読み取り成功の真偽,平均温冷感
        """        
        inst = 'analogInput:' + str(10000 * int(tenant.value) + 1000 * zone_number + self._member.ThermalSensation.value)
        return await self.read_present_value(self.target_ip,inst)
    

    async def get_averaged_clothing_index(self, tenant, zone_number):
        """ゾーンに在室している執務者数の平均着衣量を取得する
        Args:
            tenant (Tenant): テナント
            zone_number (Unsigned): ゾーン番号（1~9）
        Returns:
            list: 読み取り成功の真偽,平均着衣量
        """        
        inst = 'analogInput:' + str(10000 * int(tenant.value) + 1000 * zone_number + self._member.ClothingIndex.value)
        return await self.read_present_value(self.target_ip,inst)
    
    async def get_thermally_dissatisfied_rate(self, tenant, zone_number):
        """ゾーンに在室している執務者の温熱環境に対する不満足者率を取得する
        Args:
            tenant (Tenant): テナント
            zone_number (Unsigned): ゾーン番号（1~9）
        Returns:
            list: 読み取り成功の真偽,ゾーンに在室している執務者の温熱環境に対する不満足者率
        """        
        inst = 'analogInput:' + str(10000 * int(tenant.value) + 1000 * zone_number + self._member.Dissatisfied_Thermal.value)
        return await self.read_present_value(self.target_ip,inst)
    
    async def get_dissatisfied_rate_caused_by_draft(self, tenant, zone_number):
        """ゾーンに在室している執務者のドラフトに対する不満足者率を取得する
        Args:
            tenant (Tenant): テナント
            zone_number (Unsigned): ゾーン番号（1~9）
        Returns:
            list: 読み取り成功の真偽,ゾーンに在室している執務者のドラフトに対する不満足者率
        """        
        inst = 'analogInput:' + str(10000 * int(tenant.value) + 1000 * zone_number + self._member.Dissatisfied_Draft.value)
        return await self.read_present_value(self.target_ip,inst)
    
    async def get_dissatisfied_rate_caused_by_vertical_temperature_distribution(self, tenant, zone_number):
        """ゾーンに在室している執務者の上下温度分布に対する不満足者率を取得する
        Args:
            tenant (Tenant): テナント
            zone_number (Unsigned): ゾーン番号（1~9）
        Returns:
            list: 読み取り成功の真偽,ゾーンに在室している執務者の上下温度分布に対する不満足者率
        """        
        inst = 'analogInput:' + str(10000 * int(tenant.value) + 1000 * zone_number + self._member.Dissatisfied_VerticalTemp.value)
        return await self.read_present_value(self.target_ip,inst)

# endregion

# region 執務者別    

    async def is_occupant_stay_in_office(self, tenant, occupant_index):
        """在室しているか否かを取得する
        Args:
            tenant (Tenant): テナント
            occupant_index (Unsigned): 執務者番号（1～）
        Returns:
            list(bool,bool): 読み取り成功の真偽,在室しているか否か
        """
        inst = 'binary-input,' + str(10000 * int(tenant.value) + 10 * occupant_index + self._member.Availability.value)
        val = await self.read_present_value(self.target_ip,inst)
        return val[0], (val[1] == 1)


    async def get_thermal_sensation(self, tenant, occupant_index):
        """温冷感を取得する
        Args:
            tenant (Tenant): テナント
            occupant_index (Unsigned): 執務者番号（1～）
        Returns:
            list: 読み取り成功の真偽,温冷感
        """        
        inst = 'analogInput:' + str(10000 * int(tenant.value) + 10 * occupant_index + self._member.ThermalSensation.value)
        return await self.read_present_value(self.target_ip,inst)


    async def get_clothing_index(self, tenant, occupant_index):
        """着衣量を取得する
        Args:
            tenant (Tenant): テナント
            occupant_index (Unsigned): 執務者番号（1～）
        Returns:
            list: 読み取り成功の真偽,着衣量
        """        
        inst = 'analogInput:' + str(10000 * int(tenant.value) + 10 * occupant_index + self._member.ClothingIndex.value)
        return await self.read_present_value(self.target_ip,inst)

# endregion

async def main():
    oCom = OccupantCommunicator(15)

    while True:
        val = await oCom.get_occupant_number(OccupantCommunicator.Tenant.South)
        print('Occupant in south tenant' + (' = ' + str(val[1]) if val[0] else ' 通信失敗'))

        val = await oCom.get_occupant_number(OccupantCommunicator.Tenant.North)
        print('Occupant in north tenant' + (' = ' + str(val[1]) if val[0] else ' 通信失敗'))

        val = await oCom.is_occupant_stay_in_office(OccupantCommunicator.Tenant.South, 1)
        print('Occupant 1 stay in office' + (' = ' + str(val[1]) if val[0] else ' 通信失敗'))

        val = await oCom.get_thermal_sensation(OccupantCommunicator.Tenant.South, 1)
        print('Thermal sensation of occupant 1 ' + (' = ' + str(val[1]) if val[0] else ' 通信失敗'))

        val = await oCom.get_clothing_index(OccupantCommunicator.Tenant.South, 1)
        print('Clothing index of occupant 1 ' + (' = ' + '{:.1f}'.format(val[1]) if val[0] else ' 通信失敗'))

        time.sleep(1)

if __name__ == "__main__":
    asyncio.run(main())